### ISSUE TITLE
AllocationEvent's come before associated VTable event

### ISSUE THREAD

**Issue by rodrmoya** [mono/mono#6259]
In mono master (5.11), the MLPDs produced by the runtime profiler seem to have one problem, only happening for a few "System.RuntimeType" allocations that seem to happen at the beginning of the application. The series of events is:

(Messages obtained via adding debugging console output to Xamarin.Profiler on a v15 MLPD where the problem was seen)
Buffer ID's are the buffer's TimeBase field, and the 2nd number is the timestamp of the specific event:

1. ClassLoad event:
   The culprit is System.RuntimeType at buffer 19250786083136 at 19250788456173
2. Several Allocation events for objects of that type, without any stack trace:
   No backtrace at buffer 19250786083136 at 19250789353288
   No backtrace at buffer 19250786083136 at 19250789359463
   No backtrace at buffer 19250786083136 at 19250789362233
   No backtrace at buffer 19250786083136 at 19250789363302
   No backtrace at buffer 19250786083136 at 19250789364130
3. VTable event for the unknown-till-now VTablePointer:
   Got unknown VTable for class 2072191896 at buffer 19250786083136 at 19250789365557

All events are in the same buffer, and it is not an ordering problem because of the timestamp sorting in SyncPoint, as the events have different timestamps

**Comment by alexrp** [comment]
Here's what happens: The runtime wants to create an instance of `System.RuntimeType` (which is Mono's concrete implementation of `System.Type`) for some type. To do this, it must first fill out a `MonoVTable` for the `MonoClass` representing `System.RuntimeType`, just as with any other class being instantiated in managed code. But there's a catch: Every `MonoVTable` has a `type` field pointing to a `System.RuntimeType` instance for the `MonoClass` that `MonoVTable` corresponds to.

So, when creating the `MonoVTable` for `System.RuntimeType` itself, we have a circular dependency. To complete the loading of the `MonoVTable`, we must fill out the `type` field, but to fill out that field, we must instantiate a `System.RuntimeType` object... which needs the `MonoVTable` we're currently creating.

The runtime works around this by 'publishing' the `MonoVTable` for `System.RuntimeType` *before* setting its `type` field as a special case, which is fine. But unfortunately it's not that simple for the profiler: The loading of that `MonoVTable` can still fail *after* it's been published - namely, when the `type` field is being set. After all, allocating the `System.RuntimeType` to fill out that field could fail due to an OOM condition. So we can't just emit the `vtable_loaded` event before we set the `type` field.

TL;DR: This is basically unfixable. You have to patch up the allocations after you get the relevant VTable load event. Could you use a placeholder object/value until then? If it helps, you should only have to do this for `System.RuntimeType`; no other type has this kind of circular dependency when loading a `MonoVTable`.

**Comment by alexrp** [comment]
Discussed this with Rodrigo on Slack a while back. The Xamarin Profiler will work around this as there's not really any way for the runtime to fix it.
